### PR TITLE
Simplier alter prepare

### LINKED ISSUE #8669: Inconsistent ALTER behaviour.
**Describe the bug or unexpected behavior**
Modification of type of column with non-materialized defaults leads to wrong default values.

**How to reproduce**
```
CREATE TABLE test_alter_on_mutation
(
  date Date,
  key UInt64
)
ENGINE ReplicatedMergeTree('/clickhouse/tables/test_alter_on_mutation', '1')
ORDER BY key;

INSERT INTO test_alter_on_mutation select toDate('2020-01-05'), number from system.numbers limit 100;

ALTER TABLE test_alter_on_mutation ADD COLUMN value String DEFAULT '10';

SELECT sum(cast(value as UInt64)) from test_alter_on_mutation;

┌─sum(CAST(value, 'UInt64'))─┐
│                       1000 │
└────────────────────────────┘

-- OPTIMIZE TABLE test_alter_on_mutation FINAL;

ALTER TABLE test_alter_on_mutation MODIFY COLUMN value UInt64;

SELECT sum(value) from test_alter_on_mutation;

┌─sum(value)─┐
│          0 │
└────────────┘
```

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Backward Incompatible Change

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improved `ALTER MODIFY/ADD` queries logic. Now you cannot `ADD` column without type, `MODIFY` default expression doesn't change type of column and `MODIFY` type doesn't loose default expression value. Fixes #8669.

Documentation can be found in Changelog entry.

### CODE REVIEW

**Comment by alesapin** [comment]
Why tests for alters don't have word `alter` in their names?

File: dbms/src/Storages/AlterCommands.cpp [hunk 1]
```diff
@@ -498,189 +504,95 @@ void AlterCommands::apply(StorageInMemoryMetadata & metadata) const
 }
 
 
-void AlterCommands::prepare(const StorageInMemoryMetadata & metadata, const Context & context)
+void AlterCommands::prepare(const StorageInMemoryMetadata & metadata, const Context & /*context*/)
```
Comment (alexey-milovidov) [inline dbms/src/Storages/AlterCommands.cpp]: If it's not a virtual method or used in template - let's get rid of unused parameter completely.

File: dbms/src/Storages/AlterCommands.cpp [hunk 2]
```diff
@@ -498,189 +504,95 @@ void AlterCommands::apply(StorageInMemoryMetadata & metadata) const
...
+        const auto & column_name = command.column_name;
+        if (command.type == AlterCommand::ADD_COLUMN)
+        {
+            if (metadata.columns.has(column_name) || metadata.columns.hasNested(column_name))
             {
-                if (metadata.columns.has(column_name) || metadata.columns.hasNested(column_name))
-                    if (!command.if_not_exists)
-                        throw Exception{"Cannot add column " + column_name + ": column with this name already exists", ErrorCodes::ILLEGAL_COLUMN};
+                if (!command.if_not_exists)
+                    throw Exception{"Cannot add column " + column_name + ": column with this name already exists", ErrorCodes::ILLEGAL_COLUMN};
```
Comment (alexey-milovidov) [inline dbms/src/Storages/AlterCommands.cpp]: Let's add backQuote everywhere.

Comment (alexey-milovidov) [inline dbms/src/Storages/AlterCommands.cpp]: DUPLICATE_COLUMN

File: dbms/src/Storages/AlterCommands.cpp [hunk 3]
```diff
@@ -498,189 +504,95 @@ void AlterCommands::apply(StorageInMemoryMetadata & metadata) const
...
+                else
+                    continue;
             }
-            else if (command.type == AlterCommand::MODIFY_COLUMN)
+
+            if (!command.data_type)
+                throw Exception{"Data type have to be specified for column " + column_name + " to add", ErrorCodes::ILLEGAL_COLUMN};
```
Comment (alexey-milovidov) [inline dbms/src/Storages/AlterCommands.cpp]: BAD_ARGUMENTS

File: dbms/src/Storages/AlterCommands.cpp [hunk 4]
```diff
@@ -498,189 +504,95 @@ void AlterCommands::apply(StorageInMemoryMetadata & metadata) const
...
+        }
+        else if (command.type == AlterCommand::MODIFY_COLUMN)
+        {
+            if (!metadata.columns.has(column_name))
             {
-                if (!metadata.columns.has(column_name))
-                    if (!command.if_exists)
-                        throw Exception{"Wrong column name. Cannot find column " + column_name + " to modify", ErrorCodes::ILLEGAL_COLUMN};
+                if (!command.if_exists)
+                    throw Exception{"Wrong column name. Cannot find column " + column_name + " to modify", ErrorCodes::ILLEGAL_COLUMN};
```
Comment (alexey-milovidov) [inline dbms/src/Storages/AlterCommands.cpp]: UNKNOWN_IDENTIFIER / NOT_FOUND_COLUMN_IN_BLOCK

File: dbms/src/Storages/AlterCommands.cpp [hunk 5]
```diff
@@ -718,6 +630,30 @@ void AlterCommands::validate(const StorageInMemoryMetadata & metadata, const Con
...
+        auto copy_expression = default_expression->clone();
+        auto default_with_cast = makeASTFunction("CAST", copy_expression, std::make_shared<ASTLiteral>(column_type->getName()));
+        auto query_with_alias = setAlias(default_with_cast, tmp_column_name);
+        auto syntax_result = SyntaxAnalyzer(context).analyze(query_with_alias, all_columns.getAll());
+        ExpressionAnalyzer(query_with_alias, syntax_result, context).getActions(true);
+    }
+    catch (Exception & ex)
+    {
+        ex.addMessage("default expression and column type are incompatible. Cannot alter column '" + column_name + "'");
+        throw(ex);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/AlterCommands.cpp]: Just `throw`. It's important to keep full stack trace.

There is a difference between rethrowing and throwing a new exception.

**Pull request merged**